nemo_aligner/algorithms/ppo.py

```diff
@@ -42,6 +47,82 @@
 from nemo_aligner.utils.utils import clear_memory, cpu_dict, masked_mean
 
 
+class PPORolloutBatch(UserDict):
+    @classmethod
+    def from_rollout_batches(
+        cls, rollout_batches: List[Dict], eos_id: int, rollout_batch_seq_length: Union[int, None]
+    ):
+        """Given a list of rollout batches, stack the tensors within and put them in a single dictionary
+        """
+        stacked_dict = cls()
+
+        if len(rollout_batches) == 0:
+            return stacked_dict
+
+        keys = rollout_batches[0].keys()
+
+        for k in sorted(keys):
+
+            list_of_tensors = [item[k] for item in rollout_batches]
+
+            if all(map(lambda x: x.ndim == 1, list_of_tensors)):
+                tensor = torch.cat(list_of_tensors)
+            else:
+                pad_value = eos_id if k == "response_tokens" else 0
+
+                list_of_tensors = list(
+                    itertools.chain(*(map(lambda x: x.flatten(), item.split(1, dim=0)) for item in list_of_tensors))
+                )
+                tensor = torch.nn.utils.rnn.pad_sequence(list_of_tensors, batch_first=True, padding_value=pad_value)
+
+                # find the max sequence length globally
+                max_seqlen = torch.tensor([tensor.size(-1)], dtype=torch.long, device=torch.cuda.current_device())
+                torch.distributed.all_reduce(max_seqlen, op=torch.distributed.ReduceOp.MAX)
+
+                if rollout_batch_seq_length is None or max_seqlen > rollout_batch_seq_length:
+                    pad_seq_len = max_seqlen.item()
+                else:
+                    pad_seq_len = rollout_batch_seq_length if k == "response_tokens" else rollout_batch_seq_length - 1
+
+                tensor = torch.nn.functional.pad(tensor, (0, pad_seq_len - tensor.size(-1)), value=pad_value)
```

Comment:
I don't think this will matter in practice, if our num_rollout_samples is 512 at fp32 precision the input token tensor is 512 * 8192 * 4/1024/1024 = 16MB. We also clear all the cache and force a garbage collect before training.

Though if we had a super long context length(not sure how we can even generate that much that fast) then this might pose an issue. added a comment 8de54768aaf9460e271ce6f57829fe6e95104188

Reply:
Sounds good, thanks!